src/ray/gcs/gcs_server/gcs_resource_scheduler.cc

```diff
@@ -23,71 +23,29 @@ namespace gcs {
 
 double LeastResourceScorer::Score(const ResourceRequest &required_resources,
                                   const NodeResources &node_resources) {
-  // In GCS-based actor scheduling, the `predefined_resources` and `custom_resources` (of
-  // class `NodeResources`) are only acquired or released by actor scheduling, instead of
-  // being updated by resource reports from raylets. So we have to subtract normal task
-  // resources (if exist) from the current available resources.
+  // In GCS-based actor scheduling, the `NodeResources` are only acquired or released by
+  // actor scheduling, instead of being updated by resource reports from raylets. So we
+  // have to subtract normal task resources (if exist) from the current available
+  // resources.
   const NodeResources *node_resources_ptr = &node_resources;
   NodeResources new_node_resources;
   if (!node_resources.normal_task_resources.IsEmpty()) {
     new_node_resources = node_resources;
-    for (size_t i = 0;
-         i < node_resources.normal_task_resources.predefined_resources.size();
-         ++i) {
-      new_node_resources.predefined_resources[i].available -=
-          node_resources.normal_task_resources.predefined_resources[i];
-      if (new_node_resources.predefined_resources[i].available < 0) {
-        new_node_resources.predefined_resources[i].available = 0;
-      }
-    }
-    for (const auto &request_resource_entry :
-         node_resources.normal_task_resources.custom_resources) {
-      auto iter = new_node_resources.custom_resources.find(request_resource_entry.first);
-      if (iter != new_node_resources.custom_resources.end()) {
-        iter->second.available -= request_resource_entry.second;
-        if (iter->second.available < 0) {
-          iter->second.available = 0;
-        }
-      }
-    }
+    new_node_resources.available -= node_resources.normal_task_resources;
+    new_node_resources.available.RemoveNegative();
```

Comment:
Is it better to `RemoveNegative` in `-=` implicitly? It's more friendly for forgetful coders.

Reply:
Negative values are allowed in "ResourceRequest". For example, you may want to calculate the diff of 2 sets. So we cannot do that by default.